Add fuzzy matching

iamsims opened the pull request:
- Config use_fuzzy_match and fuzzy_match_threshold

Fuzzy match uses fuzzy token set ratio, in default fuzzy logic is set to false and the fuzzy threshold for match is 87.

iamsims committed:
Add fuzzy matching

- Config use_fuzzy_match and fuzzy_match_threshold

NISH1001 commented on akd/tools/source_validator.py:
```diff
@@ -334,18 +345,22 @@ def _validate_against_whitelist(
+                # ):
+                #     # Check if it's a meaningful match (not just common words)
+                #     if len(whitelisted_title) > 10 or len(source_title) > 10:
+                #         return True, category_name, 0.8
```
I think this remove older logic no?

Instead of having destructive change, can we have constructive change. That is: first check if fuzzy is enabled. If not enabled, then use the older logic.  Don't replace current logic because not sure fuzzy will work 100% of the time.

I'd suggest something like

```python
if self.config.use_fuzzy_match:
  ....<your code>
  ...return

continue with whatever logic we have previously. This should fix this comment.
```

NISH1001 commented on akd/tools/source_validator.py:
```diff
@@ -334,18 +345,22 @@ def _validate_against_whitelist(
+                    
+
+                if self.config.use_fuzzy_match:
+                    fuzzy_score_set = token_set_ratio(source_title, whitelisted_title)
```
Can we have the fuzzy match function name as configurable? That is `fuzzy_fn_name` or something and get the function based on that name. It will give more configurability on what fuzzy matching algo to apply

iamsims committed:
Make change to address PR comments

- Configure the type of fuzzy match function
- Fallback to the original algorithm for matching if fuzzy match is disabled

iamsims committed:
Add unit tests for source validator

NISH1001 commented on akd/tools/source_validator.py:
```diff
@@ -20,11 +21,19 @@
+class FuzzyMatchMethod(str, Enum):
+    token_set = "token_set"
+    token_sort = "token_sort"
+    ratio = "ratio"
```
Let's have enum all caps lock like `TOKEN_SET` etc....

```python
TOKEN_SET = "token_set"
...
```

NISH1001 commented on akd/tools/source_validator.py:
```diff
@@ -299,6 +325,18 @@ def _parse_crossref_response(
+        "token_set": token_set_ratio,
+        "token_sort": token_sort_ratio,
+        "ratio": ratio,
+        }
```
maybe this could be class-level attribute?

```python

class SourceValidator(...):
  ...
  _scorer_map = dict(token_set = token_set_ratio, ...)
  ...
  def __init__(self, ...):
    ...
```

iamsims committed:
Address the comments

- Change the cases of enum constants to CAPS
- Make class variable of scorer_map instead of function variable
